apps/system/js/app_chrome.js

```diff
@@ -365,22 +366,83 @@
         break;
 
       case this.pinButton:
-        this.pin();
+        this.pinSite();
         break;
     }
   };
 
+  /**
+   * Pin current page in places database.
+   */
+  AppChrome.prototype.pinPage = function ac_pinPage() {
+    Service && Service.request('Places:setPinned', this.app.config.url, true)
+      .then(function() {
+      console.log('Succeeded in pinning ' + this.app.config.url);
+    }, function() {
+      console.error('Failed to pin ' + this.app.config.url);
+    });
+  };
+
+  /**
+   * Pin current site in bookmarks database.
+   */
+  AppChrome.prototype.pinSite = function ac_pinSite() {
+    this.pin();
+
+    LazyLoader.load('shared/js/url_helper.js').then(() => {
+      var siteObject = {};
+      var manifestURL = this.app.webManifestURL;
+      var manifestObject = this.app.webManifest;
+      var pageUrl = this.app.config.url;
+
+      siteObject.type = 'url';
+      siteObject.iconable = false;
+      siteObject.icons = this.app.favicons;
+      siteObject.frecency = 1;
+      siteObject.pinned = true;
+      siteObject.pinnedFrom = pageUrl;
+
+      if (manifestURL && manifestObject) {
+        var startUrl = manifestObject.start_url ?
+          UrlHelper.resolveUrl(manifestObject.start_url, pageUrl) || pageUrl :
```

Comment:
Not sure I like this pattern... What about returning null on resolveUrl with a null start_url, instead of having ternaries here?